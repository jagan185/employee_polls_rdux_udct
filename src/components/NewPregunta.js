import { connect } from "react-redux";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import Authenticate from "./Authenticate";
import { handleAddQuestion } from "../actions/questions";

const NewPregunta = ({ dispatch, authedUser }) => {
  const navigate = useNavigate();
  const [firstOption, setFirstOption] = useState('');
  const [secondOption, setSecondOption] = useState('');
  const [success, setSuccess] = useState(false);
  const [error, setError] = useState(false);

  useEffect(() => {
    console.log({ success, error });
  }, [success, error]);

  const handleSubmit = (event) => {
    event.preventDefault();

    if (firstOption !== '' & secondOption !== '') {
      const question = {
        optionOneText: firstOption,
        optionTwoText: secondOption
      };

      dispatch(handleAddQuestion(question));
      setSuccess(true);
      setError(false);
    } else {
      setSuccess(false);
      setError(true);
    };

    setFirstOption('');
    setSecondOption('');
    navigate('/');
  };

  const optionOneChange = (event) => {
    setFirstOption(event.target.value);
  };

  const optionTwoChange = (event) => {
    setSecondOption(event.target.value);
  };

  return (
    <div className="flex flex-col max-w-screen-xl ml-5 rounded overflow-hidden shadow-lg">
      <Authenticate />
      {success && (
        <div className="m-2">
          <h1 className="text-green-900 text-xl" data-testid="success-header">New Question saved successfully</h1>
        </div>
      )}
      {error && (
        <div className="m-2">
          <h1 className="text-red-900 text-xl" data-testid="error-header"> Error saving new question</h1>
        </div>  
      )}
      <div className="m-2"><h6 className="text-black text-xl text-center font-bold">Create your own poll</h6></div>
      <div className="m-2"><h1 className="text-black text-xl text-center font-bold">Would you rather</h1> </div>
      <form className="space-y-4 md:space-y-6 m-2" onSubmit={handleSubmit}>
        <div className="m-2">
          <label htmlFor="firstOptionValue" className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
            First Option
          </label>
          <input
            value={firstOption}
            className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            name="firstOptionValue"
            placeholder="Option One"
            onChange={optionOneChange}
            data-testid="option-one-input"
            required
          />
        </div>
        <div className="m-2">
          <label htmlFor="secondOptionValue" className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
            Second option
          </label>
          <input
            value={secondOption}
            className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            name="secondOptionValue"
            placeholder="Option Two"
            onChange={optionTwoChange}
            data-testid="option-two-input"
            required
          />
        </div>
        <button
          type="submit"
          className="bg-blue-300 hover:bg-blue-700 text-black hover:text-white font-bold py-2 px-4 m-2 rounded"
          data-testid="submit-button"
        >
          Submit
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => {
  return {
    authedUser
  };
};

export default connect(mapStateToProps)(NewPregunta);