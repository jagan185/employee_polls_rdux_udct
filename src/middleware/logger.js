/**
 * logger is a function that takes store as input and returns a function that takes next as input and returns a function that takes action as input logs action and calls next function passing in action, which returns result, this result is returned by logger
 * @param {*} store 
 * @returns 
 */
const logger = (store) => (next) => (action) => {
    console.group(action.type);
    console.log("action :: ", action);
    const result = next(action);
    console.log("state from store :: ", store.getState());
    console.groupEnd();
    return result;
  };

  export default logger;