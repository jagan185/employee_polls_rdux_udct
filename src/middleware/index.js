import thunk from "redux-thunk";
import { applyMiddleware } from "redux";
import logger from "./logger.js";
/**
 * combine middlewares here,
 * middlewares are applied in the order they are passed to applyMiddleware
 */
export default applyMiddleware(thunk, logger);