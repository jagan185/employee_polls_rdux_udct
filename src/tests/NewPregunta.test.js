/**
 * @description unit tests for NewPregunta.js
 */
import { Provider } from "react-redux";
import { legacy_createStore } from "redux";
import { MemoryRouter } from "react-router-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';

import reducer from "../reducers";
import middleware from "../middleware";
import NewPregunta from "../components/NewPregunta.js";

const store = legacy_createStore(reducer, middleware);

describe("NewPregunta", () => {
  it("will save new question.", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <NewPregunta />
        </Provider>
      </MemoryRouter>
    );

    let choice1 = screen.getByTestId("choice1Txt");
    fireEvent.change(choice1, { target: { value: "Play FIFA" } });
    let choice2 = screen.getByTestId("choice2Txt");
    fireEvent.change(choice2, { target: { value: "Smoke SISHA" } });
    fireEvent.submit(choice1,choice2);
    expect(screen.getByTestId("successMsg")).toBeInTheDocument();
    expect(screen.queryByTestId("errorMsg")).not.toBeInTheDocument();
  });

  it("should match snapshot", () => {
    let { component } = render(
      <MemoryRouter>
        <Provider store={store}>
          <NewPregunta />
        </Provider>
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });
});