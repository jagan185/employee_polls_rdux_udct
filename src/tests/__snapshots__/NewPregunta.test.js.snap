// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPregunta should match snapshot 1`] = `undefined`;
